README.md
# hcolor

Simple color conversions.

# Support

- RGB
- RGBA
- HSL
- HSLA
- HEX

# Example

```js
const rgba = new HColor.RGBA(65, 43, 21, .5);
rgba.toHSLA().toString(true); // hsla(30,51%,17%,0.5)

HColor.HEX2HSLA('#135', .5, true); // hsla(210,67%,20%,0.5)

const hsl = HColor.rgb(12, 34, 56).toHSL(true);
hsl.toString(); // hsl(210,65%,13%)

HColor.parse('#F00').toString(); // rgb(255,0,0)
```

# APIs

Please read the declaration files in `typings` or the source files in `src` to learn the APIs.

# Changelog

See [CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
### 0.4.1 - 2018-08-24

- Fix a bug in `hsl.toHSL` & `hsla.toHSL`

## 0.4.0 - 2018-08-24

- Rename `StringPattern` `pattern`

## 0.3.0 - 2018-08-24

- Add `round` option to cloning methods

## 0.2.0 - 2018-08-23

- Add `parse`

### 0.1.1 - 2018-08-22

- Add `hsl.toHSL` & `hsla.toHSL`

## 0.1.0 - 2018-08-17

- Initial release


test/test.js
const HColor = require('../dist/hcolor.umd.js'),
    { strictEqual } = require('assert');

const sample = {
    rgb: 'rgb(12,34,56)',
    hsl: 'hsl(125,10%,60%)',
    hex: '#678'
};

strictEqual(
    HColor.RGB.fromString(sample.rgb).toHSL().toRGB().toString(true),
    sample.rgb
);

strictEqual(
    HColor.HSL.fromString(sample.hsl).toRGB().toHSL().toString(true),
    sample.hsl
);

strictEqual(
    HColor.HSLA.fromHEX(sample.hex, .6).toRGBA().toString(true),
    HColor.HEX2RGBA(sample.hex, .6)
);

strictEqual(
    HColor.parse('#F00').toString(),
    'rgb(255,0,0)'
);


src/HSL.ts
import { Vector } from "./Vector";
import { RGB } from "./RGB";

export class HSL extends Vector {

    constructor(h = 0, s = 0, l = 0) {
        super([h, s, l, 1]);
    }

    static fromHEX(hex: string, round = false) {
        return RGB.fromHEX(hex).toHSL(round);
    }

    static pattern = /^hsl\(\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)%\s*,\s*((?:\d*\.)?\d+)%\s*\)$/i;

    static fromString(hsl: string) {
        if (!hsl.match(HSL.pattern)) {
            throw "Invalid HSL string!";
        } else {
            return new HSL(+RegExp.$1, +RegExp.$2 / 100, +RegExp.$3 / 100);
        }
    }

    set h(value: number) {
        this.components[0] = value;
    }
    get h() {
        return this.components[0];
    }

    set s(value: number) {
        this.components[1] = value;
    }
    get s() {
        return this.components[1];
    }

    set l(value: number) {
        this.components[2] = value;
    }
    get l() {
        return this.components[2];
    }

    toString(round = false) {
        let c = this.components.slice(0, 3);
        c[1] *= 100;
        c[2] *= 100;
        if (round) {
            c = c.map(Math.round);
        }
        return `hsl(${c[0]},${c[1]}%,${c[2]}%)`;
    }

    toHSL(round = false) {
        let c = this.components.slice(0, 3);
        if (round) {
            c[1] *= 100;
            c[2] *= 100;
            c = c.map(Math.round);
            c[1] /= 100;
            c[2] /= 100;
        }
        return new HSL(c[0], c[1], c[2]);
    }

    toHSLA(alpha = 1) {
        const { components: c } = this;
        return new HSLA(c[0], c[1], c[2], alpha);
    }

    toHEX(upperCase = true) {
        return this.toRGB(true).toHEX(upperCase);
    }

    toRGB(round = false) {
        const { h, s, l } = this;
        if (s === 0) {
            return new RGB(l, l, l);
        } else {
            const q = l < .5 ? l * (1 + s) : (l + s - l * s),
                p = 2 * l - q,
                k = h / 360,
                qp6 = (q - p) * 6,
                _1_6 = 1 / 6,
                _2_3 = 2 / 3;
            let rgb = [k + 1 / 3, k, k - 1 / 3].map(
                t => t < 0 ? t + 1 : (t > 1 ? t - 1 : t)
            ).map(
                t => t < _1_6 ? p + qp6 * t : (
                    t < .5 ? q : (
                        t < _2_3 ? p + qp6 * (_2_3 - t) :
                            p
                    )
                )
            ).map(
                n => n * 255
            );
            if (round) {
                rgb = rgb.map(Math.round);
            }
            return new RGB(rgb[0], rgb[1], rgb[2]);
        }
    }

    toRGBA(alpha = 1, round = false) {
        return this.toRGB(round).toRGBA(alpha);
    }

}

// @ts-ignore
export class HSLA extends HSL {

    constructor(h = 0, s = 1, l = .5, a = 1) {
        super(h, s, l);
        this.components[3] = a;
    }

    static fromHEX(hex: string, alpha = 1, round = false) {
        return RGB.fromHEX(hex).toHSLA(alpha, round);
    }

    static pattern = /^hsla\(\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)%\s*,\s*((?:\d*\.)?\d+)%\s*,\s*((?:\d*\.)?\d+)\s*\)$/i;

    static fromString(hsla: string) {
        if (!hsla.match(HSLA.pattern)) {
            throw "Invalid HSLA string!";
        } else {
            return new HSLA(+RegExp.$1, +RegExp.$2 / 100, +RegExp.$3 / 100, +RegExp.$4);
        }
    }

    set a(value: number) {
        this.components[3] = value;
    }
    get a() {
        return this.components[3];
    }

    toString(round = false) {
        let c = this.components.slice(0);
        c[1] *= 100;
        c[2] *= 100;
        if (round) {
            c = c.slice(0, 3).map(Math.round).concat(c[3]);
        }
        return `hsla(${c[0]},${c[1]}%,${c[2]}%,${c[3]})`;
    }

    // @ts-ignore
    toRGBA(round = false) {
        return this.toRGB(round).toRGBA(this.components[3]);
    }

}


src/factories.ts
import { RGB, RGBA } from "./RGB";
import { HSL, HSLA } from "./HSL";

export const rgb = (r = 0, g = 0, b = 0) => new RGB(r, g, b);
export const rgba = (r = 0, g = 0, b = 0, a = 1) => new RGBA(r, g, b, a);
export const hsl = (h = 0, s = 1, l = .5) => new HSL(h, s, l);
export const hsla = (h = 0, s = 1, l = .5, a = 1) => new HSLA(h, s, l, a);

export const RGB2HSL = (rgb: string, round = false) => RGB.fromString(rgb).toHSL(round).toString();
export const RGB2HSLA = (rgb: string, alpha = 1, round = false) => RGB.fromString(rgb).toHSLA(alpha, round).toString();
export const RGB2HEX = (rgb: string, upperCase = true) => RGB.fromString(rgb).toHEX(upperCase);
export const RGBA2HSL = (rgba: string, round = false) => RGBA.fromString(rgba).toHSL(round).toString();
export const RGBA2HSLA = (rgba: string, round = false) => RGBA.fromString(rgba).toHSLA(round).toString();
export const RGBA2HEX = (rgba: string, upperCase = true) => RGBA.fromString(rgba).toHEX(upperCase);

export const HSL2RGB = (hsl: string, round = false) => HSL.fromString(hsl).toRGB(round).toString();
export const HSL2RGBA = (hsl: string, alpha = 1, round = false) => HSL.fromString(hsl).toRGBA(alpha, round).toString();
export const HSL2HEX = (hsl: string, upperCase = true) => HSL.fromString(hsl).toHEX(upperCase);
export const HSLA2RGB = (hsla: string, round = false) => HSLA.fromString(hsla).toRGB(round).toString();
export const HSLA2RGBA = (hsla: string, round = false) => HSLA.fromString(hsla).toRGBA(round).toString();
export const HSLA2HEX = (hsla: string, upperCase = true) => HSLA.fromString(hsla).toHEX(upperCase);

export const HEX2RGB = (hex: string) => RGB.fromHEX(hex).toString();
export const HEX2RGBA = (hex: string, alpha = 1) => RGBA.fromHEX(hex, alpha).toString();
export const HEX2HSL = (hex: string) => HSL.fromHEX(hex).toString();
export const HEX2HSLA = (hex: string, alpha = 1, round = false) => HSLA.fromHEX(hex, alpha, round).toString();


src/RGB.ts
import { Vector } from "./Vector";
import { HSL } from "./HSL";

export class RGB extends Vector {

    constructor(r = 0, g = 0, b = 0) {
        super([r, g, b, 1]);
    }

    static fromHEX(hex: string) {
        const { length } = hex;
        if (hex[0] !== '#' || length !== 4 && length !== 7) {
            throw "Invalid HEX value!";
        }
        if (length === 4) {
            return new RGB(
                parseInt(hex[1], 16) * 17,
                parseInt(hex[2], 16) * 17,
                parseInt(hex[3], 16) * 17
            );
        } else {
            return new RGB(
                parseInt(hex.slice(1, 3), 16),
                parseInt(hex.slice(3, 5), 16),
                parseInt(hex.slice(5), 16)
            );
        }
    }

    static pattern = /^rgb\(\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)\s*\)$/i;

    static fromString(rgb: string) {
        if (!rgb.match(RGB.pattern)) {
            throw "Invalid RGB string!";
        } else {
            return new RGB(+RegExp.$1, +RegExp.$2, +RegExp.$3);
        }
    }

    set r(value: number) {
        this.components[0] = value;
    }
    get r() {
        return this.components[0];
    }

    set g(value: number) {
        this.components[1] = value;
    }
    get g() {
        return this.components[1];
    }

    set b(value: number) {
        this.components[2] = value;
    }
    get b() {
        return this.components[2];
    }

    toString(round = false) {
        let c = this.components.slice(0, 3);
        if (round) {
            c = c.map(Math.round);
        }
        return `rgb(${c.join(',')})`;
    }

    toRGB(round = false) {
        let c = this.components.slice(0, 3);
        if (round) {
            c = c.map(Math.round);
        }
        return new RGB(c[0], c[1], c[2]);
    }

    toRGBA(alpha = 1) {
        return new RGBA(...this.components.slice(0, 3), alpha);
    }

    toHEX(upperCase = true): string {
        return upperCase ? this.toHEX(false).toUpperCase() :
            ('#' + this.components.slice(0, 3)
                .map(n => Math.round(n).toString(16))
                .map(s => s.length === 1 ? '0' + s : s)
                .join(''));
    }

    toHSL(round = false) {
        const [r, g, b] = this.components.slice(0, 3).map(n => n / 255),
            max = Math.max(r, g, b), min = Math.min(r, g, b),
            sum = max + min, delta = max - min,
            h = max === min ? 0 :
                max === r ? (60 * (g - b) / delta + (g < b ? 360 : 0)) :
                    max === g ? 60 * (b - r) / delta + 120 :
                        60 * (r - g) / delta + 240,
            l = sum / 2,
            s = l === 0 || max === min ? 0 :
                delta / (l > .5 ? 2 - sum : sum);
        return round ?
            new HSL(Math.round(h), Math.round(s * 100) / 100, Math.round(l * 100) / 100) :
            new HSL(h, s, l);
    }

    toHSLA(alpha = 1, round = false) {
        return this.toHSL(round).toHSLA(alpha);
    }

}

// @ts-ignore
export class RGBA extends RGB {

    constructor(r = 0, g = 0, b = 0, a = 1) {
        super(r, g, b);
        this.components[3] = a;
    }

    static fromHEX(hex: string, alpha = 1) {
        return RGB.fromHEX(hex).toRGBA(alpha);
    }

    static pattern = /^rgba\(\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)\s*,\s*((?:\d*\.)?\d+)\s*\)$/i;

    static fromString(rgba: string) {
        if (!rgba.match(RGBA.pattern)) {
            throw "Invalid RGBA string!";
        } else {
            return new RGBA(+RegExp.$1, +RegExp.$2, +RegExp.$3, +RegExp.$4);
        }
    }

    set a(value: number) {
        this.components[3] = value;
    }
    get a() {
        return this.components[3];
    }

    toString(round = false) {
        let { components: c } = this;
        if (round) {
            c = c.map((n, i) => i < 3 ? Math.round(n) : n) as typeof c;
        }
        return `rgba(${c.join(',')})`;
    }

    // @ts-ignore
    toHSLA(round = false) {
        return this.toHSL(round).toHSLA(this.components[3]);
    }

}

